internal/expr/bindinputs.go

```diff
@@ -49,41 +35,111 @@ func (te *TypedExpr) BindInputs(args ...any) (pq *PrimedQuery, err error) {
-					return nil, fmt.Errorf("parameter with type %q missing, have type with same name: %q", typeMember.OuterType().String(), t.String())
+		typeToValue[t] = v
+	}
+
+	// Generate SQL and query parameters.
+	params := []any{}
+	outputs := []typeinfo.Member{}
+	argTypeUsed := map[reflect.Type]bool{}
+	inCount := 0
+	outCount := 0
+	sqlStr := bytes.Buffer{}
+	for _, te := range *tes {
+		switch te := te.(type) {
+		case *typedInputExpr:
+			typeMember := te.input
+			outerType := typeMember.OuterType()
+			v, ok := typeToValue[outerType]
+			if !ok {
+				return nil, inputMissingError(outerType, typeToValue)
+			}
+			argTypeUsed[outerType] = true
+
+			val, err := typeMember.ValueFromOuter(v)
+			if err != nil {
+				return nil, err
+			}
+			params = append(params, sql.Named("sqlair_"+strconv.Itoa(inCount), val.Interface()))
+
+			sqlStr.WriteString("@sqlair_" + strconv.Itoa(inCount))
+			inCount++
+		case *typedOutputExpr:
+			for i, oc := range te.outputColumns {
+				sqlStr.WriteString(oc.sql)
+				sqlStr.WriteString(" AS ")
+				sqlStr.WriteString(markerName(outCount))
+				if i != len(te.outputColumns)-1 {
+					sqlStr.WriteString(", ")
 				}
+				outCount++
+				outputs = append(outputs, oc.tm)
 			}
-			return nil, fmt.Errorf("%s not referenced in query", t.Name())
+		case *bypass:
+			sqlStr.WriteString(te.chunk)
 		}
 	}
 
-	// Query parameters.
-	var params []any
-	for i, typeMember := range te.inputs {
-		outerType := typeMember.OuterType()
-		v, ok := typeValue[outerType]
-		if !ok {
-			return nil, typeMissingError(outerType.Name(), typeNames)
+	for argType := range typeToValue {
+		if !argTypeUsed[argType] {
+			return nil, fmt.Errorf("%s not referenced in query", argType.Name())
 		}
+	}
 
-		val, err := typeMember.ValueFromOuter(v)
-		if err != nil {
-			return nil, err
-		}
+	return &PrimedQuery{outputs: outputs, sql: sqlStr.String(), params: params}, nil
+}
 
-		params = append(params, sql.Named("sqlair_"+strconv.Itoa(i), val.Interface()))
+// typedExpression represents a expression bound to a type. It contains
+// information to generate the SQL for the part and to access Go types
+// referenced in the part.
+type typedExpression interface {
+	// typedExpr is a marker method.
+	typedExpr()
+}
+
+// outputColumn stores the name of a column to fetch from the database and the
+// type member to scan the result into.
+type outputColumn struct {
+	sql string
+	tm  typeinfo.Member
+}
+
+// typedOutputExpr contains the columns to fetch from the database and
+// information about the Go values to read the query results into.
+type typedOutputExpr struct {
+	outputColumns []outputColumn
+}
+
+// typedExpr is a marker method.
+func (*typedOutputExpr) typedExpr() {}
+
+// checkUsed checks if the members in the typed output expression have already
+// been used in the query. It updates the memberUsed map with its own members.
+func (toe *typedOutputExpr) checkUsed(memberUsed map[typeinfo.Member]bool) error {
```

Comment:
That's fair, I'm also not sure about it being a method on typedOutputExpr. I added this function because I didn't like that we were digging into the typedOutputExpr in BindTypes, if we update the map there we still have to do that really. I just inline this function again and changed the map name to remove the reference to member.